src/main/java/org/opensearch/security/resources/ResourceAccessHandler.java

```diff
@@ -388,12 +401,66 @@ public void revoke(
     }
 
     /**
-     * Loads all resources within the specified resource index.
+     * Loads all resource-ids within the specified resource index.
      *
      * @param resourceIndex The resource index to load resources from.
      * @param listener      The listener to be notified with the set of resource IDs.
      */
-    private void loadAllResources(String resourceIndex, ActionListener<Set<String>> listener) {
+    private void loadAllResourceIds(String resourceIndex, ActionListener<Set<String>> listener) {
         this.resourceSharingIndexHandler.fetchAllResourceIds(resourceIndex, listener);
     }
+
+    /**
+     * Loads all resource-sharing records for the specified resource index.
+     *
+     * @param resourceIndex The resource index to load records from.
+     * @param listener      The listener to be notified with the set of resource-sharing records.
+     */
+    private void loadAllResourceSharingRecords(String resourceIndex, ActionListener<Set<SharingRecord>> listener) {
+        this.resourceSharingIndexHandler.fetchAllResourceSharingRecords(resourceIndex, listener);
+    }
+
+    /**
+     * Returns flat principals to be used when querying the sharing index and while searching resource-ids.
+     * @param user user whose security-config (name, roles and backend_roles) is to be flattened.
+     * @return the set of flattened principals
+     */
+    private Set<String> getFlatPrincipals(User user) {
+        // 1) collect all entities we’ll match against share_with arrays
+        // for users:
+        Set<String> users = new HashSet<>();
+        users.add(user.getName());
+        users.add("*"); // for matching against publicly shared resource
+
+        // for roles:
+        Set<String> roles = new HashSet<>(user.getSecurityRoles());
+        roles.add("*"); // for matching against publicly shared resource
```

Comment:
Let's use the convention of `*` only in users for publicly visible. This could even be a separate field to indicate if the resource is sharable which would improve searchability of these. 

In https://github.com/opensearch-project/security/pull/5600, there's a `all_shared_principals` field added to the resource doc itself (outside sharing index) and it would contain `user:*` as the convention for meaning the resource was publicly shared.